status-im/status-react
# Refactor Quo preview

### Summary

In the spirit of recent improvements made to the Quo preview main screen, this PR brings the following:

- Change `quo-preview.preview` code to be more aligned with our guidelines.
- Remove duplication when setting up preview screens in `quo-preview.main`. Now we only need to specify the `:name` and `:component` keys and fallback to a good default for `{:options {:topBar {:visible true}}`.
- Auto-generate descriptor labels based on the key.
- Fix form field colors for dark & light, especially for Android.
- Redesigned form fields to look nicer 💄
- Create component that abstracts away the code for rendering the form fields and the preview area. This will aid us in creating more consistent looking preview screens and keep things DRY.

### How do we use the new code?

Just as before, define a `state` binding and use it to build up the arguments of the component. Then, use the `preview/preview-container` to wrap the component. You can use certain props to control how the preview is styled, etc, but that's about it.

```clojure
(def descriptor
  [{:type :text :key :label}
   {:key     :chevron-position
    :type    :select
    :options [{:key :left}
              {:key :right}]}])

(defn view
  []
  (let [state (reagent/atom {:chevron-position :left
                             :label            "Welcome"})]
    (fn []
      [preview/preview-container {:state state :descriptor descriptor}
       [quo/divider-label @state]])))
```

### So how are we going to update the rest of the preview screens?

It's fairly easy to refactor them, I just didn't do in this PR because the point was just to show that the abstraction works across a variety of different previews (including a few with blur).

I think we don't need to update guideline docs now. If we just update a bunch of previews in a follow up PR, then devs will start to pick up the pattern.

### Demo

<table>
  <thead>
    <tr>
      <th>Before</th>
      <th>After</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><img src="https://github.com/status-im/status-mobile/assets/46027/8631cde6-8627-4562-a20f-8c4f2208b7f0" width="350"/></td>
      <td><img src="https://github.com/status-im/status-mobile/assets/46027/a8eb0431-23b2-44c8-8400-592a0ea473c9" width="350"/></td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th>Before</th>
      <th>After</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><img src="https://github.com/status-im/status-mobile/assets/46027/056cb419-5e75-48c6-be47-eebd3e8bf1fe" width="350"/></td>
      <td><img src="https://github.com/status-im/status-mobile/assets/46027/f6e8ae2a-ca59-4b65-90a9-393a554a5d3b" width="350" /></td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th>Before</th>
      <th>After</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><img src="https://github.com/status-im/status-mobile/assets/46027/7ccc0535-36db-4899-a6e0-0f695527a48b" width="350"/></td>
      <td><img src="https://github.com/status-im/status-mobile/assets/46027/6229cd1c-bccc-4ef2-a30f-ca107e312a89" width="350"/></td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th>Before</th>
      <th>After</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><img src="https://github.com/status-im/status-mobile/assets/46027/749b96a3-da70-42bf-9e0f-1d3e05348469" width="350"/></td>
      <td><img src="https://github.com/status-im/status-mobile/assets/46027/e780dbf5-042c-4978-8800-57b01f7fe827" width="350"/></td>
    </tr>
  </tbody>
</table>

### Review notes

- I have updated a few preview screens to help you see the boilerplate going away. Nothing changed behavior-wise.
- From now on (future PRs) it will be very cheap to fix other preview screens to use the new `preview-container` component and remove all that copy & paste of `cool-preview`, `flat-list`, and so on.

### Areas that maybe impacted

No impact whatsoever.

### Steps to test

Check any of the preview screens to see the pigs with new lipstick. Or also specifically check the preview screens that this PR refactored to use the new `preview-container` component, spoiler, you won't see any meaningful difference, so don't bother :)

status: ready
